File: backend/src/main/java/com/projectps/cinema/service/MovieService.java
package com.projectps.cinema.service;

import com.projectps.cinema.DTO.MovieDTO;
import com.projectps.cinema.entity.Movie;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public interface MovieService {

    Movie saveMovie(MovieDTO movieDTO);

    List<Movie> saveMovies(List<MovieDTO> moviesDTO);

    List<MovieDTO> getMovies();

    MovieDTO getMovieById(int id);

    List<MovieDTO> getMoviesByGenre(String genre);

    List<MovieDTO> getMoviesByScore(double score);

    List<MovieDTO> getMoviesByYear(int year);

    Movie updateMovie(MovieDTO movieDTO);

    void deleteMovie(int id);

    List<MovieDTO> sortMoviesByTitle(String order);

    List<MovieDTO> sortMoviesByScore(String order);

    MovieDTO incresePopularity(int id);

    MovieDTO decreasePopularity(int id);

    List<MovieDTO> getMoviesByPopularity();
}

File: frontend/app/service/Movie.service.ts
import { Injectable } from "@angular/core";
import {HttpClient} from '@angular/common/http';
import { Movie } from "../model/Movie";


@Injectable({
  providedIn: "root"
})
export class MovieService {

  baseUrl: string = "http://localhost:8081/movie";

  constructor(private httpClient:HttpClient) {
  }

  addMovie(movie: Movie) {
    return this.httpClient.post(this.baseUrl + "/addMovie", movie);
  }

  addMovies(movies: Movie[]) {
    return this.httpClient.post(this.baseUrl + "/addMovies", movies);
  }

  getAllMovies() {
    return this.httpClient.get<Movie[]>(this.baseUrl + "/allMovies");
  }

  getMovieById(id: number) {
    return this.httpClient.get<Movie>(this.baseUrl + "/byId/" + id);
  }

  getMoviesByGenre(genre: string) {
    return this.httpClient.get<Movie[]>(this.baseUrl + "/byGenre/" + genre);
  }

  getMoviesByScore(score: number) {
    return this.httpClient.get<Movie[]>(this.baseUrl + "/byScore/" + score);
  }

  getMoviesByYear(year: number) {
    return this.httpClient.get<Movie[]>(this.baseUrl + "/byYear/" + year);
  }

  updateMovie(movie: Movie) {
    return this.httpClient.put(this.baseUrl + "/updateMovie", movie);
  }

  deleteMovie(id: number) {
    return this.httpClient.delete(this.baseUrl + "/deleteMovie/" + id);
  }

  sortMoviesByTitle(order: string) {
    return this.httpClient.get<Movie[]>(this.baseUrl + "/sortMoviesByTitle/" + order);
  }

  sortMoviesByScore(order: string) {
    return this.httpClient.get<Movie[]>(this.baseUrl + "/sortMoviesByScore/" + order);
  }

  sortMoviesByPopularity() {
    return this.httpClient.get<Movie[]>(this.baseUrl + "/sortMoviesByPopularity");
  }

}
